File: leetcode/chain/82.cpp
#include <iostream>

#include "list_node.h"

class Solution {
 public:
  ListNode* delete_duplicates(ListNode* node) {
    ListNode *last = new ListNode(0, node);
    ListNode *ret = last;
    while (node != nullptr) { // 遍历
      bool is_duplicates = false;
      while (node->next != nullptr && node->val == node->next->val) { //删除重复节点
        node->next = node->next->next;
        is_duplicates = true;
      }
      if (is_duplicates) { // 当前节点之前存在重复节点
        node = node->next;
        last->next = node;
      } else {
        last = node;
        node = node->next;
      }
    }
    return ret->next;
  }
};

int main() {
  ListNode* node = new ListNode(1);
  node->append(new ListNode(3));
  node->append(new ListNode(3));
  node->append(new ListNode(4));
  node->append(new ListNode(4));
  node->append(new ListNode(5));
  node->print();

  Solution solution = Solution();
  ListNode* ret = solution.delete_duplicates(node);
  ret->print();
}

File: leetcode/chain/82.py
# Definition for singly-linked list.
class ListNode:

    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:

    def deleteDuplicates(self, head: ListNode) -> ListNode:
        '''删除排序链表中的重复元素 II'''
        ret = ListNode()
        last = ret
        while head is not None:
            flag = False  #当前位置是否删除了相同节点
            while head.next is not None and head.val == head.next.val:  #删除相同的节点
                head.next = head.next.next
                flag = True
            if flag:
                head = head.next
                last.next = head
            else:
                last.next = head
                last = last.next  #挪位，只有此处进行last的挪位
                head = head.next
        return ret.next
